### ISSUE TITLE
Efficient inference methods for RNN

### ISSUE THREAD

**Issue by ZachL1** [openvinotoolkit/openvino#11156]
Hello!

### Description
I have a recurrent neural network that I need to process every frame of a video. The output of the previous inference will be used as the input of the next inference, how should I use OpenVINO to achieve the **best efficiency**?

### Detail
This is what I am doing now. But I think this may not be efficient, the get tensor and then set tensor seem a bit redundant.
```cpp
infer_request.set_tensor(input_name, init_tensor);
for (frame : video) {
    // set tensor for frame
    // ......
    infer_request.infer();
    auto out_tensor = infer_request.get_tensor(out_name);
    infer_request.set_tensor(input_name, out_tensor);
}
```

Can these be implemented efficiently inside OpenVINO? Or any better way than this?

**Comment by itikhono** [comment]
Hi @ZachL1
 Yes, there is a way to automate this process, OpenVINO has special Assign/ReadValue operations, which allow you to store a Tensor between inferences. This mechanism works on GNA and CPU plugins.
 
![image](https://user-images.githubusercontent.com/21181928/159700882-9f45bbb0-0f86-499f-aa4a-391fd218baf9.png)

 In the next release we will introduce a transformation to quickly replace your Parameter/Result pair with ReadValue/Assign operations. MakeStateful transformation is already available in master branch, so if you use master, you can call it from c++ code
 
```
    ov::Core core;
    auto ov_model = core.read_model("path_to_the_model");
    // Parameter_1, Result_1, Parameter_3, Result_3 are shared_ptr<Parameter/Result> in the ov_model
    std::vector<std::pair<std::shared_ptr<ov::opset8::Parameter>, std::shared_ptr<ov::opset8::Result>>> pairs
            = {/*Parameter_1, Result_1, Parameter_3, Result_3*/};
    ov::pass::Manager manager;
    manager.register_pass<ov::pass::MakeStateful>(pairs); // you can also use tensor names here of Parameter/Result nodes
    manager.run_passes(ov_model);

```
 Or when converting the model via ModelOptimizer
` --input_model <INPUT_MODEL> --transform "MakeStateful[param_res_names={'tensor_name_1':'tensor_name_4','tensor_name_3':'tensor_name_6'}]"`

But if I remember correctly, we have some performance issue on CPU, maybe not with all models (@vladislav-volkov please comment) , so I suggest just trying and comparing with the original.

**Comment by ZachL1** [comment]
@itikhono Thank you very much! This is exactly what I need and I will try your suggestion asap.

<br>
Allow me to ask another question I just ran into, I'm not sure if I should create a new issue, I'll leave it here for now.

### Detail
I'm converting ONNX to IR and I want to use both NCHW->NHWC and BGR->RGB:
```bash
mo --input_model matting_onnx_model.onnx --input src,r1i,r2i,r3i,r4i --output pha,r1o,r2o,r3o,r4o --layout "src(nchw->nhwc),r1i(n???),r2i(n???),r3i(n???),r4i(n???)" --reverse_input_channels
```
However I got the error:
```bash
[ ERROR ]  Network has 5 inputs overall, but none of them are suitable for input channels reversing.
Suitable for input channel reversing inputs are 4-dimensional with 3 channels (in case of dynamic dimensions C channel must be provided in a layout for this input)
All inputs: [['src', <PartialShape: {1,3,1080,1920}>], ['r1i', <PartialShape: {1,16,135,240}>], ['r2i', <PartialShape: {1,20,68,120}>], ['r3i', <PartialShape: {1,40,34,60}>], ['r4i', <PartialShape: {1,64,17,30}>]]
```
This is strange because the `src` input has a dimension that is 3-channel.

If I try them separately:
```bash
mo --input_model matting_onnx_model.onnx --input src,r1i,r2i,r3i,r4i --output pha,r1o,r2o,r3o,r4o --layout "src(nchw->nhwc),r1i(n???),r2i(n???),r3i(n???),r4i(n???)"
```
```bash
mo --input_model matting_onnx_model.onnx --input src,r1i,r2i,r3i,r4i --output pha,r1o,r2o,r3o,r4o --reverse_input_channels
```
They all worked.

Is this a bug? I can't apply NCHW->NHWC and BGR->RGB at the same time.

**Comment by itikhono** [comment]
@ZachL1 yes, it looks like a bug
@mvafin @rkazants is this a known issue? I can't find a ticket

**Comment by ZachL1** [comment]
> Or when converting the model via ModelOptimizer ` --input_model <INPUT_MODEL> --transform "MakeStateful[param_res_names={'tensor_name_1':'tensor_name_4','tensor_name_3':'tensor_name_6'}]"`
> 
> But if I remember correctly, we have some performance issue on CPU, maybe not with all models (@vladislav-volkov please comment) , so I suggest just trying and comparing with the original.

Excellent work! I've tried using it when converting models, and I ended up with models that will no longer need input to hide state, which is great! It works fine on my CPU now, but I haven't tested its performance in detail yet, I'm going to do that.

In addition, the converted model does not allow me to input the hidden state again, how does OpenVINO set the initial value of the hidden state? Is it a random value or all zeros?

**Comment by itikhono** [comment]
by default they are zeros, but you can change it using OV State API.
StateAPI methods:
```
 * `std::string get_name() const`
   returns name(variable_id) of according Variable
 * `void reset()`
   reset state to default value
 * `void set_state(const Tensor& state)`
   set new value for state
 * `Tensor get_state() const`
   returns current value of state
```
   
   Example:
  ```
 CompiledModel hw_specific_model = ie.compile_model(network, "CPU");

    // Create Infer Request
    InferRequest inferRequest = hw_specific_model.create_infer_request();

    // Reset to zeros all memory states before starting
    auto states = inferRequest.query_state();
    for (auto&& state : states) {
        state.reset();
    }
    
    // Set a new value to state
    std::vector<float> v = {5};
    Tensor tensor(element::f32, Shape{1, 1});
    std::memcpy(tensor.data(), &v[0], sizeof(float));
    states[0].set_state(tensor);
```

**Comment by mvafin** [comment]
> Is this a bug? I can't apply NCHW->NHWC and BGR->RGB at the same time.

Hi @ZachL1 

Are you still experiencing the issue with nchw->nhwc and --reverse_input_channels? I noticed you created an issue about it and closed it.

**Comment by ZachL1** [comment]
> by default they are zeros, but you can change it using OV State API. StateAPI methods:

@itikhono I think all my doubts are solved, thank you!
In addition, I just compared in detail the model with Assign/ReadValue operations and the model without it. I used CPU inference. assign/ReadValue operations does seem to bring a slight, though not significant, performance improvement. But anyway, Assign/ReadValue operations at least made my code writing clean, which is nice. Once again, great job!

<br>

> Are you still experiencing the issue with nchw->nhwc and --reverse_input_channels? I noticed you created an issue about it and closed it.

@mvafin Yes, it is still a problem. I think this may be a bug. You can reproduce it with this [ONNX model](https://drive.google.com/file/d/10IzwKCPLX72HVlSMTLELGbp3-JjLnSjC/view?usp=sharing):
```
pip install openvino-dev==2022.1.0
mo --input_model omz_matting.onnx --layout "img(nchw->nhwc),s1i(n???),s2i(n???),s3i(n???),s4i(n???)" --reverse_input_channels
```

**Comment by mvafin** [comment]
Thank you @ZachL1. I can reproduce the issue, I will update you when the fix is prepared.

**Comment by jgespino** [comment]
@ZachL1 Closing issue, please test using the latest master branch and start a new issue if you run into any issues.
